src/packbuilder.rs

```diff
@@ -0,0 +1,398 @@
+
+    /// Recursively insert an object and its referenced objects. Insert the
+    /// object as well as any object it references.
+    pub fn insert_recursive(&mut self,
+                            id: Oid,
+                            name: Option<&str>)
+                            -> Result<(), Error> {
+        let name = try!(::opt_cstr(name));
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_recur(self.raw,
+                                                        &*id.raw(),
+                                                        name));
+        }
+        Ok(())
+    }
+
+    /// Write the contents of the packfile to an in-memory buffer. The contents
+    /// of the buffer will become a valid packfile, even though there will be
+    /// no attached index.
+    pub fn write_buf(&mut self, buf: &mut Buf) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_write_buf(buf.raw(), self.raw));
+        }
+        Ok(())
+    }
+
+    /// Create the new pack and pass each object to the callback.
+    pub fn foreach<F>(&mut self, mut cb: F) -> Result<(), Error>
+        where F: FnMut(&[u8]) -> bool {
+        let mut cb = &mut cb as &mut ForEachCb;
+        let ptr = &mut cb as *mut _;
+        unsafe {
+            try_call!(raw::git_packbuilder_foreach(self.raw,
+                                                   foreach_c,
+                                                   ptr as *mut _));
+        }
+        Ok(())
+    }
+
+    /// `progress` will be called with progress information during pack
+    /// building. Be aware that this is called inline with pack building
+    /// operations, so performance may be affected.
+    ///
+    /// There can only be one progress callback attached, this will replace any
+    /// existing one. See `unset_progress_callback` to remove the current
+    /// progress callback without attaching a new one.
+    pub fn set_progress_callback<F>(&mut self, progress: F) -> Result<(), Error>
+        where F: FnMut(PackBuilderStage, u32, u32) -> bool + 'repo {
+        let ptr = Box::into_raw(Box::new(Box::new(progress) as Box<ProgressCb>));
+        let progress_c = Some(progress_c as raw::git_packbuilder_progress);
+        let ret;
+        unsafe {
+            ret = ::call::try(call!(
+                    raw::git_packbuilder_set_callbacks(self.raw,
+                                                       progress_c,
+                                                       ptr as *mut _)));
+            // Current implementation will never return an error, but since it
+            // may change in the future...
+            // If an error is returned we don't know if it has kept the old
+            // pointers or the new ones. So to avoid dropping a callback that
+            // may actually get called in the future we keep the old callback
+            // stored and just leak the new one by not recreating a `Box` from
+            // the raw pointer.
```

Comment:
That's the semantics I'd expect as well, but they don't say what the error semantics are in the header. Currently they only return an error if the packbuilder pointer is null, which we ensure never happens; but if it changes in the future for some reason I'd prefer to have the current wrapper memory safe with a tiny performance/leak penalty.

Is there any other way to get the raw pointer of a box without using `into_raw`? The only alternative I can think of would be something along the lines of

```rust
let progress = Box::new(progress) as Box<ProgressCb>;
let ptr = &mut progress as *mut _;
unsafe {
    raw::git_packbuilder_set_callbacks(self.raw, progress_c, ptr as *mut _);
    self.progress = progress;
}
```

but that would be unsafe as the pointer would be to the stack allocation of the box before it got moved into the `Packbuilder`.